### DISCUSSION TITLE
Understanding the use case for text button

### DISCUSSION THREAD

**Discussion by joaomarcelofm** [primefaces/community#2006] [Figma UI Kit]
Hello everyone!

Our team is finding it hard to understand the use case for the `text` variant of buttons. The main issue is that most of the time, it requires custom styling to prevent the UI from looking misaligned.

For instance, in the [sample code](https://stackblitz.com/edit/gogsjj?file=src%2FApp.vue):
<img width="509" alt="Screenshot 2024-05-23 at 07 36 04" src="https://github.com/primefaces/community/assets/30631810/0661e281-d162-4595-b9eb-1e05f6fc5d6b">

In the example above, the UI looks buggy and misaligned until the user hovers and sees that some background is applied to the padding. This requires us to apply custom styles to make it look as our designers desire, but generally, how text buttons look everywhere, I think 🤔
Also, even if in the "good" example it looks aligned, then when the user hovers, it'd look strange if we used `pl-0` as one of the sides wouldn't have a background. Therefore, we often remove that style as well.
<img width="587" alt="Screenshot 2024-05-23 at 08 00 47" src="https://github.com/primefaces/community/assets/30631810/f2945610-b93c-4c18-a4a8-3f14dadbc3b4">

Based on [other discussions](https://github.com/orgs/primefaces/discussions/28), this seems to be a common issue. Do you have any guidelines for best practices regarding the usage of this button?

Our team is resorting to the full ban on this button style as the custom styling hampers our ability to upgrade prime easily or introduces unintended bugs in the UI.

I understand the need for extra padding for icon-only buttons, as it increases the touch area for accessibility. Still, it generally requires a custom implementation to make it aligned (i.e., dynamically position the button beyond the box boundaries). This may be less of an issue, but it is also important to think about.
<img width="232" alt="Screenshot 2024-05-23 at 07 45 14" src="https://github.com/primefaces/community/assets/30631810/b13dff4b-ec48-4245-b658-968288571ba7">

**Comment by onursenture** [comment]
Hello, 

I understand the problems with the screenshots you've provided. However, if you have concerns about this alignment, they are likely due to misusing a text button. 

Personally, I wouldn't recommend removing the left padding just to achieve left alignment when the button is not in a hover state. In such cases, you should avoid using a text button. That's why there are numerous variations of the button component available. 

For instance, you can try the outlined button or the link button, which already has no padding.

**Reply by joaomarcelofm** [reply]
Thanks for the response @onursenture!

> if you have concerns about this alignment, they are likely due to misusing a text button.

That's precisely what I'm trying to understand. Is it expected that a text button will never be around the edges of a component?

However, the expectation of text buttons not being around edges, and therefore breaking alignment, is also broken in some prime elements. For instance, the dialog element:

![image](https://github.com/primefaces/community/assets/30631810/0dd16103-7c4a-4566-8169-7d5f2a4134f2)

This might be a design decision, which is fair. But the alignment of the button also breaks seemingly unintentionally in cases where the test of the dialog header is larger.

![image](https://github.com/primefaces/community/assets/30631810/048e8513-878e-4f95-97dd-352621b79f3f)

In the case above, I'd expect the button to be placed relatively to the box, not the sibling content. That's one of the issues with the current text button.
Also, I'm not saying the text button doesn't have a use case, but there are many cases where the default usage makes the UI looks buggy.

Airbnb has many examples of using text buttons whose content is aligned to the rest of the content instead of its outer padding.

![image](https://github.com/primefaces/community/assets/30631810/8acff102-c1a1-41a5-beff-d0751de6a079)

> Personally, I wouldn't recommend removing the left padding just to achieve left alignment when the button is not in a hover state.

I totally agree here. We aren't doing it, but as shown in the linked discussion there are folks that may be resorting to it in order to make their UI look aligned.

> In such cases, you should avoid using a text button. That's why there are numerous variations of the button component available.
For instance, you can try the outlined button or the link button, which already has no padding.

This isn't a solution for us. The text button is relevant for [low emphasis](https://m3.material.io/components/buttons/guidelines#c9bcbc0b-ee05-45ad-8e80-e814ae919fbb) actions and applying a different button style may completely change the meaning of it.

While researching this, I came to two possible solutions:
- Apply negative margins to push the buttons to the correct placement. This is how Airbnb is doing it.

<img width="811" alt="Screenshot 2024-05-26 at 10 15 32" src="https://github.com/primefaces/community/assets/30631810/4eed5725-6e44-4d2f-a65e-c42e23f00db9">

- (Preferred) Apply the hover effect as a pseudo-element and let only the button's content take up box-size space (no padding). This is how Google does it:

![image](https://github.com/primefaces/community/assets/30631810/dcaa486d-a8dc-428c-aaad-3eada1cb8880)

It'd be neat if this were a component prop, something like: `boxSize: boolean`. `false` would remove the padding and apply the styles as pseudo elements. The first option is already achievable today but requires a lot more custom styling.

\* Used Airbnb as an example, as they are a company recognized for following and implementing great design principles.

**Reply by onursenture** [reply]
We can't say that text buttons should never be used in corners. We also have such uses in our components. The 'perfect square' situation you mentioned should be managed by theming or changing the component structure at the use site rather than altering the button component itself.

I will discuss this with my developer colleagues. Thank you for bringing this issue to our attention and suggesting solutions.

**Reply by joaomarcelofm** [reply]
Thanks for the response @onursenture, I really appreciate it!

> The 'perfect square' situation you mentioned should be managed by theming or changing the component structure at the use site rather than altering the button component itself.

Indeed, that was a divergence from the topic and only somewhat related. I'll play with the theme to see if we can achieve that result.
